src/h3/stream.rs

```diff
@@ -370,51 +332,130 @@ impl Stream {
 
                     _ => (),
                 }
-
-                self.state = State::FramePayloadLenLen;
             },
 
             _ => return Err(Error::UnexpectedFrame),
         }
 
         self.frame_type = Some(ty);
 
+        self.state_transition(State::FramePayloadLen, 1);
+
         Ok(())
     }
 
-    pub fn parse_frame(&mut self) -> Result<()> {
-        // Parse the whole frame payload but only if there is enough data in
-        // the stream buffer. stream.buf_bytes() returns an error if we don't
-        // have enough.
-        if self.frame_type.is_none() {
-            return Err(Error::InternalError);
+    /// Sets the frame's payload length and transitions to the next state.
+    pub fn set_frame_payload_len(&mut self, len: u64) -> Result<()> {
+        assert_eq!(self.state, State::FramePayloadLen);
+
+        // Only expect frames on Control, Request and Push streams.
+        if self.ty == Some(Type::Control) ||
+            self.ty == Some(Type::Request) ||
+            self.ty == Some(Type::Push)
+        {
+            self.state_transition(State::FramePayload, len as usize);
+
+            return Ok(());
         }
 
+        Err(Error::InternalError)
+    }
+
+    /// Tries to fill the state buffer by reading data from the corresponding
+    /// transport stream.
+    ///
+    /// When not enugh data can be read to complete the state, this returns
+    /// `Error::Done`.
+    pub fn try_fill_buffer(
+        &mut self, conn: &mut crate::Connection,
+    ) -> Result<()> {
+        let buf = &mut self.state_buf[self.state_off..self.state_len];
+
+        let (read, _) = conn.stream_recv(self.id, buf)?;
+
+        trace!(
+            "{} read {} bytes on stream {}",
+            conn.trace_id(),
+            read,
+            self.id,
+        );
+
+        self.state_off += read;
+
+        if !self.state_buffer_complete() {
+            return Err(Error::Done);
+        }
+
+        Ok(())
+    }
+
+    /// Tries to fill the state buffer by reading data from the given cursor.
+    ///
+    /// This is intended to replace `try_fill_buffer()` in tests, in order to
+    /// avoid having to setup a transport connection.
+    #[cfg(test)]
+    fn try_fill_buffer_for_tests(
```

Comment:
In some other files, I've seen these test utility functions placed closer to the tests. I don't mind either way but pointing it out in case you meant to move it.

Reply:
The difference is that this is a `Connection` method, so it needs to be inside the `impl` block. It might be possible to make this a normal function and move inside test `tests` module though.